### PR TITLE
Minimal working example complete with Neuroconv & Pynapple integration

### PR DESCRIPTION
**Pull request by Samvitti**
**Summary of Changes**
Created `workflow.py`: a script that runs the minimal working example. It converts a Pickle file into a nwb file with Neuroconv, loads data from the nwb file as Pynapple objects, runs the ISI decoder training loop with the Pynapple objects (and with cross-fold validation), runs a testing loop on the decoder and evaluates accuracy, and then saves the decoder as a Pickle file. 
Created `pickle_data_interface.py, pickle_schema.json, pickle_to_nwb.py` for NeuroConv & Pynapple IO functionality. 
Modified `session.py, n_decoder.py` for major debugging and for allowing integration with Pynapple objects instead of numpy arrays. Also added new functions to generate stratified folds for cross-fold validation, evaluate accuracy, and save the decoder as a Pickle file. 
Modified `unique_interval_set.py`'s padding functionality to mimic Pynapple's IntervalSet API even when padding is added, without needing to generate a new UniqueIntervalSet object every time the user wants to work with padded intervals.
Slightly modified `event_set.py` to improve time complexity. 

All other modifications are minor stylistic/error-handling additions. 

**Testing**
Ran `python3 workflow.py`. This script currently uses `AE_238_1_AC.pickle` as the train data and `AE_238_2_AC.pickle` as the test data. 
Generated decoder has a mean validation accuracy of 0.539906, test accuracy of 0.564417.

### CODE REVIEW

File: decoder/core/event_set.py [hunk 1]
```diff
@@ -2,7 +2,7 @@ class EventSet:
     """Represents a collection of events in a session.
 
     Attributes:
-        possible_values (list): List of possible labels/values for events.
+        possible_values (set): Set of possible labels/values for events.
```
Comment (badralbanna) [inline decoder/core/event_set.py:5]: This reminds me that we should probably be adding [typing annotations](https://docs.python.org/3/library/typing.html). Have you worked with these before in python? Maybe we can start by adding them to new functions and put this as a task on the backlog. (Also occurs to me that we should maybe start using github's "issues" tab as a work tracker)

File: decoder/core/event_set.py [hunk 2]
```diff
@@ -60,7 +63,7 @@ def delete_event(self, timepoint):
             raise KeyError(f"Timepoint {timepoint} does not exist.")
         del self.events[timepoint]
 
-    def get_sorted_events(self):
+    def get_sorted_timestamps(self):
```
Comment (badralbanna) [inline decoder/core/event_set.py:66]: Nice catch.

File: decoder/core/ndecoder.py [hunk 3]
```diff
@@ -0,0 +1,309 @@
+import numpy as np 
+import pandas as pd 
+from KDEpy import FFTKDE 
+from scipy.interpolate import interp1d 
+from sklearn.model_selection import StratifiedKFold
+from decoder.core.model import Model
+import pickle
+        
+class NDecoder:
```
Comment (badralbanna) [inline decoder/core/ndecoder.py:9]: Nit: don't worry about this now (another one for the backlog), but we should change the name of this class (probably the repo too). I know you didn't come up with this but "Neural Decoder" is too generic - there are many, many other neural decoding approaches and we should give ours a name that clearly identifies the method being used.

File: decoder/core/ndecoder.py [hunk 4]
```diff
@@ -0,0 +1,309 @@
...
+            bandwidth (float): The bandwidth parameter for the KDE.
+
+        Returns:
+            tuple: A tuple containing two functions:
+                - A function that interpolates the estimated probability density function (PDF).
+                - A function to generate an inverse Cumulative Distribution Function (CDF) for sampling.
+        """
+        # x = np.linspace(-2, 6, 100)
+        # TODO: write x in relation to fftkde data's min and max
+        x = np.linspace(-2, 20, 100)
```
Comment (badralbanna) [inline decoder/core/ndecoder.py:52]: How do FFTKDE and interp1d handle values outside the data range?

File: decoder/core/ndecoder.py [hunk 5]
```diff
@@ -0,0 +1,309 @@
...
+                Predicted probabilities for each condition for each trial.
+
+        """
+        probabilities = []
+        for trial_logISIs in X:
+            _, _, trial_probs, _ = self._predict_single_trial(self.model, trial_logISIs)
+            probabilities.append([trial_probs[cond] for cond in self.conditions])
+        return pd.DataFrame(probabilities, columns=self.conditions)
+
+    # TODO: ask if this is necessary to implement
```
Comment (badralbanna) [inline decoder/core/ndecoder.py:222]: No, I don't think this is necessary, and if it is I wouldn't handle it in this class. I would let the user sample the data as they like and then run the algorithm.

File: decoder/core/ndecoder.py [hunk 6]
```diff
@@ -0,0 +1,309 @@
...
+        # y_nogo_idx = np.where(y_nogo_mask)[0]
+        # idx_go_new = np.random.choice(y_go_idx,ymin,replace=False)
+        # idx_nogo_new = np.random.choice(y_nogo_idx,ymin,replace=False)              #Needs to stay in order
+        # idx_new = np.concatenate((idx_go_new,idx_nogo_new))
+        # X = X[idx_new]
+        # y = y[idx_new]
+
+        pass
+
+    def generate_stratified_K_folds(self, X, y, K):
```
Comment (badralbanna) [inline decoder/core/ndecoder.py:254]: Do you think this is something we need to include? The user can presumably rely on the same class we are using here. LMK if you disagree, and you think it is useful as a helper function.

File: decoder/core/ndecoder.py [hunk 7]
```diff
@@ -0,0 +1,309 @@
...
+        for train_idx, validate_idx in skf.split(X, y):
+            train_X = X[train_idx]
+            train_y = y[train_idx]
+            validate_X = X[validate_idx]
+            validate_y = y[validate_idx]
+
+            train_validate_pairs.append(((train_X, train_y), (validate_X, validate_y)))
+        return train_validate_pairs
+
+    def calculateAccuracy(self, test_X, test_y):
```
Comment (badralbanna) [inline decoder/core/ndecoder.py:272]: Same point as with stratified k folds. Do you think we need to include this or should we just output the predictions and targets in a way that can be easily fed into the metrics from sklearn? I do think we should show how to use those metrics in an example or script.

File: decoder/core/ndecoder.py [hunk 8]
```diff
@@ -0,0 +1,309 @@
...
+            # return (num_correct/num_preds),(weighted_correct/num_preds),(num_emptyISIs/num_preds)
+            return (num_correct/num_preds), (num_empty_ISIs/num_preds)
+    
+    def save_as_pickle(self, filename):
+        file = open(filename, 'wb')
+        pickle.dump(self, file)
+        file.close()
+
+
+    def fit_transform(self, X):
```
Comment (badralbanna) [inline decoder/core/ndecoder.py:306]: Don't think we need this in our case. Maybe just raise a `NotImplementedError` with an explanation

File: decoder/io/pickle_data_interface.py [hunk 9]
```diff
@@ -0,0 +1,220 @@
...
+        '''
+        Pickle file format:
+
+            data = pickle.load(file)
+            spike_train = data.spikes.times # ndarray
+            starts = data.trials.starts # ndarray
+            ends = data.trials.ends # ndarray
+
+            incorporate later if needed?:
+            lick_times = data.behavior.lick_times # ndarray
```
Comment (badralbanna) [inline decoder/io/pickle_data_interface.py:60]: This seems too specfiic, no? lick time is just an event like any other that we can use to define intervals etc...

File: decoder/io/pickle_to_nwb.py [hunk 10]
```diff
@@ -0,0 +1,19 @@
...
+import pynapple as nap 
+from decoder.io.pickle_data_interface import PickleDataInterface
+
+def neuroconv_to_nwb(interface, save_path):
+    metadata = interface.get_metadata()
+    print("Metadata extracted")
+    print(metadata)
+
+    # Add the time zone information to the conversion
+    session_start_time = datetime(2024, 6, 17, 12, 30, 0, tzinfo=ZoneInfo("EST"))
```
Comment (badralbanna) [inline decoder/io/pickle_to_nwb.py:12]: Did you mean to hardcode in the start_time here?

File: decoder/core/ndecoder.py (hunk 6 above)
Comment (badralbanna) [reply decoder/core/ndecoder.py:254]: Scratch that, I missed where you used this below. I think this makes sense, but I would make it a static method

File: decoder/workflow.py [hunk 11]
```diff
@@ -0,0 +1,206 @@
...
+from core.ndecoder import NDecoder
+from core.session import Session
+from core.bandwidth import Bandwidth
+from core.unique_interval_set import UniqueIntervalSet
+import pickle 
+import numpy as np
+import pynapple as nap 
+from decoder.io.pickle_to_nwb import pickle_to_nwb
+
+def test_loading_pickle(file_name, interval_set_name):
```
Comment (badralbanna) [inline decoder/workflow.py:11]: These are great. I'm assuming in the future we'll break these out into a testing folder for proper unit tests?

**Review by badralbanna — APPROVED** [review]
Looks great overall! Thanks for all your hard work. Feel free to respond to any comments here or in Slack.

File: decoder/AE_238_1_AC.pickle
[the remark this answers is not in the document]
Comment (Samvitti) [reply decoder/AE_238_1_AC.pickle:1]: Yup, I've used git lfs before, just forgot to use it this time. Will keep this in mind for the future!

File: decoder/core/event_set.py (hunk 1 above)
Comment (Samvitti) [reply decoder/core/event_set.py:5]: Agreed! I've worked with type annotations before. I can add them all in before my next PR; it shouldn't take me very long. Some of the classes & functions are also missing comments so I'll add those in too.

Comment (badralbanna) [reply decoder/core/event_set.py:5]: Sounds good. Yes please doument the reminaing functions.

File: decoder/core/ndecoder.py (hunk 3 above)
Comment (Samvitti) [reply decoder/core/ndecoder.py:9]: Added to Issues tab.

File: decoder/core/ndecoder.py (hunk 7 above)
Comment (badralbanna) [reply decoder/core/ndecoder.py:272]: Saw that later on you use this function. Feel free to leave and maybe make it a static method.

File: decoder/core/ndecoder.py (hunk 4 above)
Comment (Samvitti) [reply decoder/core/ndecoder.py:52]: In FFTKDE's evaluate function, it raises a ValueError if any points passed in are outside of the data range (the error message is "Every data point must be inside of the grid."). 
interp1d also throws a ValueError, but there is a parameter `bounds_error` that you can set to False so that interp1d just sets the out-of-range values with some `fill_value` (NaN by default) instead of throwing an error.

File: decoder/io/pickle_to_nwb.py (hunk 10 above)
Comment (Samvitti) [reply decoder/io/pickle_to_nwb.py:12]: I hardcoded it in just so I could quickly test if Neuroconv works. Just changed this to get the session date from the pickle file instead. The session start time is a required metadata field for NWB files so I have to pass something in to get Neuroconv to work.

File: decoder/workflow.py (hunk 11 above)
Comment (Samvitti) [reply decoder/workflow.py:11]: Yup, sounds good!

File: decoder/io/pickle_data_interface.py (hunk 9 above)
Comment (Samvitti) [reply decoder/io/pickle_data_interface.py:60]: Agreed. For now, I'll read in the lick times stored in the pickle as the event times, but in the future we may want to allow the user to pass in a pickle_read_file function that specifies how their pickle file is structured and what fields they want to load in as the event times.

File: decoder/core/ndecoder.py (hunk 3 above)
Comment (badralbanna) [reply decoder/core/ndecoder.py:10]: Great, thanks.

File: decoder/core/ndecoder.py (hunk 4 above)
Comment (badralbanna) [reply decoder/core/ndecoder.py:103]: Hmm, I think we want the range to have some padding that goes beyond the data range (KDE should naturally be able to handle values outside the observed range). That said, we should let it throw and error and handle that error in the decoding step.

File: decoder/io/pickle_data_interface.py (hunk 9 above)
Comment (badralbanna) [reply decoder/io/pickle_data_interface.py]: That's a good idea. Please add to the issues as well.

File: decoder/io/pickle_to_nwb.py (hunk 10 above)
Comment (badralbanna) [reply decoder/io/pickle_to_nwb.py]: Great. SG.
